prov/gni/src/gnix_vc.c

```diff
@@ -0,0 +1,1156 @@
+	 * earlier for this conn req.
+	 */
+	len = _gnix_dgram_unpack_buf(dgram,
+				     GNIX_DGRAM_OUT_BUF,
+				     &peer_smsg_attr,
+				     sizeof(gni_smsg_attr_t));
+
+	assert(len == sizeof(gni_smsg_attr_t));
+
+	/*
+	 *  now build the SMSG connection
+	 */
+	status = GNI_EpCreate(ep->nic->gni_nic_hndl,
+			     ep->nic->tx_cq,
+			     &vc->gni_ep);
+	if (status != GNI_RC_SUCCESS) {
+		ret = gnixu_to_fi_errno(status);
+		goto err;
+	}
+
+	status = GNI_EpBind(vc->gni_ep,
+			    peer_address.device_addr,
+			    peer_address.cdm_id);
+	if (status != GNI_RC_SUCCESS) {
+		ret = gnixu_to_fi_errno(status);
+		goto err1;
+	}
+
+	status = GNI_SmsgInit(vc->gni_ep,
+			      &local_smsg_attr,
+			      &peer_smsg_attr);
+	if (status != GNI_RC_SUCCESS) {
+		ret = gnixu_to_fi_errno(status);
+		goto err1;
+	}
+
+	status = GNI_EpSetEventData(vc->gni_ep,
+				    local_id,
+				    peer_id);
+	if (status != GNI_RC_SUCCESS) {
+		ret = gnixu_to_fi_errno(status);
+		goto err1;
+	}
+
+	/*
+	 * transition the VC to connected
+	 * put in to the nic's work queue for
+	 * further processing
+	 */
+
+	vc->conn_state = GNIX_VC_CONNECTED;
+	ret = _gnix_dgram_free(dgram);
+	if (ret != FI_SUCCESS)
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "gnix_mbox_free returned %d\n", ret);
+	vc->dgram = NULL;
+
+	ret = _gnix_vc_add_to_wq(vc);
+	if (ret == FI_SUCCESS)
+		ret = _gnix_nic_progress(nic);
+
+	return ret;
+err1:
+	GNI_EpDestroy(vc->gni_ep);
+err:
+	vc->conn_state = GNIX_VC_CONN_ERROR;
+	return ret;
+}
+
+/*
+ * One of my wildcards matched an incoming connection
+ * request from peer.  In this case we may have a vc for this
+ * peer in the hash table of vc's for this ep.
+ */
+
+static int __gnix_vc_hndl_wc_match_con(struct gnix_datagram *dgram,
+					struct gnix_address peer_address)
+{
+	int ret = FI_SUCCESS;
+	struct gnix_vc *vc = NULL, *wc_vc = NULL;
+	struct gnix_fid_ep *ep;
+	struct gnix_nic *nic;
+	int local_id, peer_id;
+	gni_smsg_attr_t local_smsg_attr;
+	gni_smsg_attr_t peer_smsg_attr;
+	gni_return_t __attribute__((unused)) status;
+	gnix_ht_key_t key;
+	ssize_t __attribute__((unused)) len;
+
+	GNIX_TRACE(FI_LOG_EP_CTRL, "\n");
+
+	wc_vc = (struct gnix_vc *)dgram->cache;
+	assert(vc);
+
+	ep = wc_vc->ep;
+	assert(ep);
+
+	nic = ep->nic;
+	assert(nic);
+
+	/*
+	 * should be in connecting state at this
+	 * point.
+	 */
+	assert(wc_vc->conn_state == GNIX_VC_CONNECTING);
+
+	/*
+	 * if the wc vc is not in the hash table
+	 * that means we are hitting the case where
+	 * we were trying to connect with a peer
+	 * who was also connecting to us, and
+	 * so hit the GNI_RC_ERROR_RESOURCE case
+	 * when posting a datagram.  We need to
+	 * move any pending sends on to the wc_vc,
+	 * remove the current vc for this peer from
+	 * the hash table, and insert the wc_vc
+	 */
+	if (!(wc_vc->modes & GNIX_VC_MODE_IN_HT)) {
+		memcpy(&key, &peer_address,
+			sizeof(gnix_ht_key_t));
+		vc = gnix_ht_lookup(ep->vc_ht, key);
+		assert(vc != NULL);
+		ret = gnix_ht_remove(ep->vc_ht, key);
```

Comment:
Do you care about the return value here (and in the lines below)?  They're getting dropped on the floor now.